router/pkg/connectrpc/handler.go

```diff
@@ -0,0 +1,665 @@
+	case http.StatusConflict: // 409
+		return connect.CodeAborted
+	case http.StatusPreconditionFailed: // 412
+		return connect.CodeFailedPrecondition
+	case http.StatusRequestEntityTooLarge: // 413
+		return connect.CodeResourceExhausted
+	case http.StatusRequestedRangeNotSatisfiable: // 416
+		return connect.CodeOutOfRange
+	case http.StatusTooManyRequests: // 429
+		return connect.CodeResourceExhausted
+	case http.StatusRequestTimeout: // 408
+		return connect.CodeDeadlineExceeded
+	case http.StatusGatewayTimeout: // 504
+		return connect.CodeDeadlineExceeded
+	case http.StatusNotImplemented: // 501
+		return connect.CodeUnimplemented
+	case http.StatusServiceUnavailable: // 503
+		return connect.CodeUnavailable
+	case http.StatusInternalServerError: // 500
+		return connect.CodeInternal
+	default:
+		// For any other status code (including 2xx success codes),
+		// return CodeUnknown as a safe default
+		return connect.CodeUnknown
+	}
+}
+
+var (
+	ErrInternalServer = errors.New("internal server error")
+)
+
+// requestHeadersKey is a custom context key for storing request headers
+type requestHeadersKey struct{}
+
+// withRequestHeaders adds request headers to the context
+func withRequestHeaders(ctx context.Context, headers http.Header) context.Context {
+	return context.WithValue(ctx, requestHeadersKey{}, headers)
+}
+
+// headersFromContext extracts the request headers from the context
+func headersFromContext(ctx context.Context) (http.Header, error) {
+	headers, ok := ctx.Value(requestHeadersKey{}).(http.Header)
+	if !ok {
+		return nil, fmt.Errorf("missing request headers")
+	}
+	return headers, nil
+}
+
+// Metadata keys for Connect error metadata
+const (
+	MetaKeyHTTPStatus          = "http-status"
+	MetaKeyErrorClassification = "error-classification"
+	MetaKeyGraphQLErrors       = "graphql-errors"
+	MetaKeyGraphQLPartialData  = "graphql-partial-data"
+	MetaKeyHTTPResponseBody    = "http-response-body"
+)
+
+// Error classification values
+const (
+	ErrorClassificationCritical = "CRITICAL"
+	ErrorClassificationPartial  = "PARTIAL"
+)
+
+// GraphQLRequest represents a GraphQL request structure
+type GraphQLRequest struct {
+	Query     string          `json:"query"`
+	Variables json.RawMessage `json:"variables,omitempty"`
+}
+
+// GraphQLErrorLocation represents the location of an error in the GraphQL query
+type GraphQLErrorLocation struct {
+	Line   int `json:"line"`
+	Column int `json:"column"`
+}
+
+// GraphQLError represents an error returned in a GraphQL response
+type GraphQLError struct {
+	Message    string                 `json:"message"`
+	Path       []any                  `json:"path,omitempty"`
+	Locations  []GraphQLErrorLocation `json:"locations,omitempty"`
+	Extensions map[string]any         `json:"extensions,omitempty"`
+}
+
+// GraphQLResponse represents a GraphQL response structure
+type GraphQLResponse struct {
+	Errors []GraphQLError  `json:"errors,omitempty"`
+	Data   json.RawMessage `json:"data,omitempty"`
+}
```

Comment:
Those types (at least part of them) already exist in the `mcpserver` package. Wouldn't it make sense to move them to a shared package and re-use them instead of defining them again?